### PR TITLE
fix: edge case fix on iceberg target table being dropped instead on being renamed first

### PR DESCRIPTION
**Pull request by nicor88**
# Description

See https://getdbt.slack.com/archives/C013MLFR7BQ/p1717497450266499

Pretty much https://github.com/dbt-athena/dbt-athena/pull/625 introduced a regression that was not letting iceberg working in an full high available mode.
This PR cover the issue, and add functional tests to catch these edge case.
Plus it introduce some improvements that allow to work properly in some restrictive LF permission models, where drop can happen only on existing objects.

## Models used to test - Optional
See functional tests

## Checklist

- [ ] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [ ] You kept your Pull Request small and focused on a single feature or bug fix.
- [ ] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

File: dbt/include/athena/macros/materializations/models/table/table.sql [hunk 1]
```diff
@@ -133,20 +133,25 @@
...
         {{ rename_relation(tmp_relation, target_relation) }}
 
-        -- old_bkp_relation might not exists in case we have a switch from hive to iceberg
-        -- we prevent to drop something that doesn't exist even if drop_relation is able to deal with not existing tables
-        {%- if old_bkp_relation is not none -%}
-          {%- do drop_relation(old_bkp_relation) -%}
+        -- old_relation_bkp might not exists in case we have a switch from hive to iceberg
+        -- if we are here old_relation_bkp was created, so we can drop it
+
+        {%- if old_relation_bkp is not none -%}
```
Comment (nicor88) [inline dbt/include/athena/macros/materializations/models/table/table.sql:151]: here I have to revert back to use old_relation_bkp and not old_bkp_relation (that is gett via a get_relation call), because old_bkp_relation seems to be always None, even if it's not....

**Comment by nicor88** [comment]
@octiva might give some input here, as he has a LF env with restricted policies, and we want to be sure that all works fine.

File: dbt/include/athena/macros/materializations/models/table/table.sql [hunk 2]
```diff
@@ -133,20 +133,27 @@
...
           {%- do drop_relation_glue(old_relation) -%}
         {%- endif -%}
 
         {{ rename_relation(tmp_relation, target_relation) }}
 
-        -- old_bkp_relation might not exists in case we have a switch from hive to iceberg
-        -- we prevent to drop something that doesn't exist even if drop_relation is able to deal with not existing tables
-        {%- if old_bkp_relation is not none -%}
-          {%- do drop_relation(old_bkp_relation) -%}
+        -- old_relation_bkp might not exists in case we have a switch from hive to iceberg
```
Comment (Avinash-1394) [inline dbt/include/athena/macros/materializations/models/table/table.sql:148]: If we are sure that the target relation has been updated at this point why not simply issue a `DROP IF EXISTS` and avoid this call altogether?

Comment (spencer-horton-wcq) [reply dbt/include/athena/macros/materializations/models/table/table.sql:148]: @Avinash-1394, that would occur if the `native_drop`is set, but wouldnt work for direct glue calls

Comment (octiva) [reply dbt/include/athena/macros/materializations/models/table/table.sql:148]: @Avinash-1394, that would occur if the native_dropis set, but wouldnt work for direct glue calls

File: dbt/include/athena/macros/materializations/models/table/table.sql [hunk 3]
```diff
@@ -133,20 +133,27 @@
...
 
         {%- set old_relation_table_type = adapter.get_glue_table_type(old_relation) -%}
+        -- we cannot use old_bkp_relation, because it returns None if the relation doesn't exist
+        -- we need to create a python object via the make_temp_relation instead
+        {%- set old_relation_bkp = make_temp_relation(old_relation, '__bkp') -%}
 
-        {%- if old_relation_table_type == 'iceberg' -%}
-          {{ rename_relation(old_relation, old_bkp_relation) }}
+        {%- if old_relation_table_type.value == 'iceberg_table' -%}
+          {{ rename_relation(old_relation, old_relation_bkp) }}
```
Comment (octiva) [inline dbt/include/athena/macros/materializations/models/table/table.sql:141]: has it been considered, that the rename might have never occurred, as the old relation value was not being correctly compared, and so, it was always None?

**Review by octiva — COMMENTED** [review]
@nicor88 I reviewed my previous work, and my changes were in earlier parts of this file: #550. 

I think the suggestion ive made about the rename not occurring ever could resolve? Keen to work it out.

File: dbt/include/athena/macros/materializations/models/table/table.sql [hunk 4]
```diff
@@ -133,20 +133,27 @@
...
 
-        -- old_bkp_relation might not exists in case we have a switch from hive to iceberg
-        -- we prevent to drop something that doesn't exist even if drop_relation is able to deal with not existing tables
-        {%- if old_bkp_relation is not none -%}
-          {%- do drop_relation(old_bkp_relation) -%}
+        -- old_relation_bkp might not exists in case we have a switch from hive to iceberg
+        -- we cannot use old_bkp_relation, because result could be cached by dbt, we use instead glue apis
+        -- get_glue_table returns none in case EntityNotFoundException, therefore this works well with restrictive environment
+        -- where Lakeformation is used, and allow operations in only existing objects
+        {%- set old_relation_bkp_exists = adapter.get_glue_table(old_relation_bkp) -%}
```
Comment (octiva) [inline dbt/include/athena/macros/materializations/models/table/table.sql:152]: We will get a 403 permission denied code here, as the table doesnt exist, which is different to the EntityNotFoundException. 😭

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 2 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/table.sql:148]: Just to make clear why we have 2 behaviours for dropping, when you run `DROP IF EXISTS` for iceberg tables, the location is left dirty in case of big tables, therefore our drop relation implementation uses glue apis with s3 apis to drop a table.

Also there is an edge case where we switch from hive to iceberg, where old_bkp_relation doesn't not exist.

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 3 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/table.sql:141]: relation type cannot be None. here you are in a part of the code where old_relation definitely exists AND we need to get the type of relation.
If the old relation is an external table == hive, we don't rename, because hive doesn't support renaming, but if the old relation is iceberg, we can simply rename the old relation to `old_relation_bkp`.

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 4 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/table.sql:152]: we have 2 options: 
1. we modify get_glue_table to tackle the exception that you have, and in that case return None
2. we create a does_table_exists, that is a wrapper on top of get_glue_table, and it returns True only if the table exists, otherwise it fallback to False, and should cover your edge case.

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 3 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/table.sql:141]: @octiva indeed, what you raised is actually a good idea to check.
the renaming happen only from old target being iceberg, for hive to iceberg old_relation_bkp doesn't exists for sure.

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 4 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/table.sql]: I addressed this comment, please have a look.

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 3 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/table/table.sql]: @octiva I should have addressed your concern here adding a check if old_relation exists, only then calling get_glue_table_type

**Comment by nicor88** [comment]
@octiva it will be good if you can test this feature again in your environment to see if all works as expected.
Thanks.

@Avinash-1394 I hope that I cover and addressed your question/suggestion. Feel free to have a look.

File: dbt/include/athena/macros/materializations/models/table/table.sql [hunk 5]
```diff
@@ -132,21 +132,29 @@
...
         {{ rename_relation(tmp_relation, target_relation) }}
 
-        -- old_bkp_relation might not exists in case we have a switch from hive to iceberg
-        -- we prevent to drop something that doesn't exist even if drop_relation is able to deal with not existing tables
-        {%- if old_bkp_relation is not none -%}
-          {%- do drop_relation(old_bkp_relation) -%}
+        -- if old relation is iceberg_table, we have a backup
+        -- therefore we can drop the old relation backup, in all other cases there is nothing to do
+        -- in case of switch from hive to iceberg the backup table do not exists
+        -- in case of fist run the backup table do not exists
```
Comment (octiva) [inline dbt/include/athena/macros/materializations/models/table/table.sql:153]:
```suggestion
        -- in case of first run, the backup table do not exists
```

**Review by Jrmyy — APPROVED** [review]
